rust/lance-index/src/traits.rs

```diff
@@ -0,0 +1,84 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use async_trait::async_trait;
+
+use lance_core::{format::Index, Result};
+
+use crate::{IndexParams, IndexType};
+
+// Extends Lance Dataset with secondary index.
+///
+#[async_trait]
+pub trait DatasetIndexExt {
+    /// Create indices on columns.
+    ///
+    /// Upon finish, a new dataset version is generated.
+    ///
+    /// Parameters:
+    ///
+    ///  - `columns`: the columns to build the indices on.
+    ///  - `index_type`: specify [`IndexType`].
+    ///  - `name`: optional index name. Must be unique in the dataset.
+    ///            if not provided, it will auto-generate one.
+    ///  - `params`: index parameters.
+    ///  - `replace`: replace the existing index if it exists.
+    async fn create_index(
+        &mut self,
+        columns: &[&str],
+        index_type: IndexType,
+        name: Option<String>,
+        params: &dyn IndexParams,
+        replace: bool,
+    ) -> Result<()>;
+
+    /// Read all indices of this Dataset version.
+    async fn load_indices(&self) -> Result<Vec<Index>>;
+
+    /// Loads all the indies of a given UUID.
+    ///
+    /// Note that it is possible to have multiple indices with the same UUID,
+    /// as they are the deltas of the same index.
+    async fn load_index(&self, uuid: &str) -> Result<Option<Index>> {
+        self.load_indices()
+            .await
+            .map(|indices| indices.into_iter().find(|idx| idx.uuid.to_string() == uuid))
+    }
+
+    /// Loads a specific index with the given index name
+    async fn load_index_by_name(&self, name: &str) -> Result<Option<Index>> {
+        self.load_indices()
+            .await
+            .map(|indices| indices.into_iter().find(|idx| idx.name == name))
+    }
+
+    /// Loads a specific index with the given index name.
+    async fn load_scalar_index_for_column(&self, col: &str) -> Result<Option<Index>>;
+
+    /// Optimize indices.
+    async fn optimize_indices(&mut self) -> Result<()>;
+
+    /// Find index with a given index_name and return its serialized statistics.
+    async fn index_statistics(&self, index_name: &str) -> Result<Option<String>>;
+
+    /// Count the rows that are not indexed by the given index.
+    ///
+    /// TODO: move to [DatasetInternalExt]
+    async fn count_unindexed_rows(&self, index_name: &str) -> Result<Option<usize>>;
+
+    /// Count the rows that are indexed by the given index.
+    ///
+    /// TODO: move to [DatasetInternalExt]
+    async fn count_indexed_rows(&self, index_name: &str) -> Result<Option<usize>>;
```

Comment:
All of these methods return `Result<Option<...>>` which is a little confusing.  We should document what `None` means in each case because it is slightly different depending on the function.  Or, we could probably change most of these to just `Result<...>` and raise an error in the `None` case.

```
load_index_by_name -> No index exists with that name, maybe just error here?
load_scalar_index_for_column -> No index exists for that column, should definitely error here looking at usage
index_statistics -> No index exists with that name, we should error here
count_unindexed_rows -> An index exists, but we couldn't determine the row count because of old manifest version
count_indexed_rows -> An index exists, but we couldn't determine the row count because of old manifest version
```

For example, looking at the python version of `count_unindexed_rows`, it is wrong:

```
    fn count_unindexed_rows(&self, index_name: String) -> PyResult<Option<usize>> {
        let idx = RT.block_on(None, self.ds.load_index_by_name(index_name.as_str()))?;
        if let Some(index) = idx {
            RT.block_on(
                None,
                self.ds
                    .count_unindexed_rows(index.uuid.to_string().as_str()),
            )?
            .map_err(|err| PyIOError::new_err(err.to_string()))
        } else {
            THIS IS NOT THE CORRECT ERROR MESSAGE
            Err(PyIOError::new_err(format!(
                "Index {} not found",
                index_name
            )))
        }
    }
```

Reply:
This was copied from the previous code. I meant to fix them in the next follow up.

Also try to eliminate these single used APIs and just put them into the json blob in `index_stats`.